### PR TITLE
Refactoring low-level data processing

### PR DESCRIPTION
**Pull request by co9olguy**

### CODE REVIEW

File: parser.py [hunk 1]
```diff
@@ -1,240 +1,18 @@
-from collections import namedtuple
+from enum import Enum
 import sympy
 
 from lark import Lark, Transformer
 
+from parser_utils import *
```
Comment (co9olguy) [inline parser.py:6]: Moved all these here

File: parser.py [hunk 2]
```diff
@@ -421,7 +159,7 @@ def goplist(self, *args):
         if len(args) == 1:
             # <uop>
             # <term>
-            return NamedList("goplist", flatten(args))
+            return ParsedList(Lists.GOPLIST, flatten(args))
```
Comment (co9olguy) [inline parser.py:162]: Felt that enums were a better method to represent these various list structures. By the time they appear in the Transformer, the type of list is more or less just for bookkeeping for the grammar, and has no real consequence

File: parser.py [hunk 3]
```diff
@@ -510,13 +250,16 @@ def uop(self, *args):
...
+            #  "simultaneous" gates/channels via a tensor product
+            op = Op(Ops.OPERATOR, op_name, params, wires)
             return TensorOp(op, other_op)
 
-        if extra_args[0].name == "idlist":
+        if extra_args[0].name == Lists.IDLIST:
             # TODO: this case should not occur, according
             # to the original spec. Determine what is causing it
-            # @co9olguy: this is because self.gatedecl calls
+            # @josh146: this is because self.gatedecl calls
```
Comment (co9olguy) [inline parser.py:262]: don't want this comment to be attributed to me :laughing:

File: parser_utils.py [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+    BARRIER = 4
+
+
+class Declarations(Enum):
+    """Enum class for representing various declarations that are allowed by the grammar."""
+    QREG = 0
+    CREG = 1
+    GATE = 2
+    OPERATOR = 3
+    CHANNEL = 4
```
Comment (co9olguy) [inline parser_utils.py:27]: I figured this kind of structure would make things more easily extendible than empty subclasses

File: parser_utils.py [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+
+
+class ParsedList:
```
Comment (co9olguy) [inline parser_utils.py:30]: This class is new, replaces the old `NamedList`

File: parser_utils.py [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+    if isinstance(obj, (list, tuple)):
+        flat_list = []
+        for item in obj:
+            flat_list.extend(flatten(item))
+        return flat_list
+    elif isinstance(obj, ParsedList):
+        return flatten(obj.list)
+    else:
+        return [obj]
+
```
Comment (co9olguy) [inline parser_utils.py:57]: Removed the other `flatten_recursive_list` function that was only used in two places, and made the `flatten` function compatible with what that deleted function was doing (by changing input to `*args`)

File: parser_utils.py [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+        if self.func == "pow":
+            return "{} ^ {}".format(*self.args)
+
+        return "{}({}, {})".format(*self.tuple)
+
+
+class Op:
+    """Class for representing quantum operations found in the grammar."""
+    def __init__(self, kind, name, params, wires):
+        self.kind = kind
```
Comment (co9olguy) [inline parser_utils.py:160]: Added the `kind` attribute (which points to an enum)

File: parser_utils.py [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+class Gate(Op):
+    """Class for representing unitary quantum gates."""
+    def __init__(self, name, params, wires):
+        super().__init__(Ops.GATE, name, params, wires)
+
+
+class Channel(Op):
+    """Class for representing non-unitary quantum channels."""
+    def __init__(self, name, params, wires):
+        super().__init__(Ops.CHANNEL, name, params, wires)
```
Comment (co9olguy) [inline parser_utils.py:181]: There are no examples that use this, but it's trivial to add to the grammar (same syntax as gates

File: parser_utils.py [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+
+    For example, as in ``qreg q[5];``"""
+    def __init__(self, name, size):
+        super().__init__(decl_type=Declarations.QREG, id_=name, size=size)
+
+    def __repr__(self):
+        return f"qreg {self.kwargs['id_']}[{self.kwargs['size']}];"
+
+
+class GateDeclaration(Declaration):
```
Comment (co9olguy) [inline parser_utils.py:312]: Removed `QDeclaration`, which wasn't doing anything useful

File: parser.py (hunk 1 above)
Comment (josh146) [reply parser.py:6]: Will need to remove the `*` import when we add linting 😆

File: parser.py [hunk 10]
```diff
@@ -255,51 +33,12 @@ def __repr__(self):
...
-
-    def declaration_str(self):
-        return f"operator {self.kwargs['op']}"[:-1]
-
-
 class QASMToIRTransformer(Transformer):
+    """Transformer for processing the Lark parse tree.
 
+    Transformers visit each node of the tree, and run the appropriate method on it according to the node's data.
+    All method names mirror the corresponding symbols from the grammar.
```
Comment (josh146) [inline parser.py:40]: We may as well update the symbols from the grammar, it is very dense and hard to read (e.g., things like `uop` and `goplist` should have better names when lexing (such as `unary_op`), it's a big overhead for new developers)

File: parser.py [hunk 11]
```diff
@@ -328,7 +67,7 @@ def statement(self, *args):
             # <decl>
             decl = args[0]
 
-            if isinstance(decl, QDeclaration):
+            if decl.name in ['gate', 'operator', 'channel']:
```
Comment (josh146) [inline parser.py:70]: 💯 🙏

File: parser.py (hunk 2 above)
Comment (josh146) [reply parser.py:162]: 🙏 💯

File: parser.py [hunk 12]
```diff
@@ -479,14 +217,16 @@ def uop(self, *args):
 
         if op_name == "CX":
             params = [None]
-            wires = [x.list for x in extra_args]
+            wires = flatten(extra_args)
```
Comment (josh146) [inline parser.py:220]: I wouldn't mind if we could also remove the recursive flattening. It seems uneeded, I'm sure we could find a workaround

File: parser.py (hunk 3 above)
Comment (josh146) [reply parser.py:262]: 😆 this comment block reminds me of
image

File: parser.py [hunk 13]
```diff
@@ -510,13 +250,16 @@ def uop(self, *args):
             # <id> <anylist> , <uop>
             # <id>() <anylist>, <uop>
             # <id>(<explist>) <anylist>, <uop>
-            op = Op(op_name, params, wires)
+            # This case only appears for tensor products of operators
+            # TODO: consider extending the grammar so users can specify
+            #  "simultaneous" gates/channels via a tensor product
```
Comment (josh146) [inline parser.py:255]: I wouldn't mind keeping simultaneous gates/channels as they are, e.g.,
```c
z q[2];
s q[2];
h q[2];
```

Using the tensor product as well here add complexity, since there is now two different ways of specifying the same thing:

```c
z q[2] @ s q[2] @ h q[2];
```

(assuming that's what you mean by simultaneous gate applications).

Comment (josh146) [reply parser.py:255]: P.s. we still need to get around to implementing https://github.com/co9olguy/qdata/commit/e5b89849c17c96705ee3888e108104afcafef0b2#r42899786. This is a almost trivial change in the parser (`,`->`@`), but I find this a _massive_ readability improvement over CSV

File: parser_utils.py [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+    GATE = 2
+    OPERATOR = 3
+    CHANNEL = 4
+
+
+class ParsedList:
+    """Bookkeeping class for the various list-like objects encountered while parsing."""
+    def __init__(self, name, list_):
+        self.name = name
+        self.list = list_
```
Comment (josh146) [inline parser_utils.py:34]: Question: why is this needed? I don't quite see it...

File: parser_utils.py [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+
+    def __repr__(self):
+        return "<{}: {}>".format(self.name, self.list)
+
+
+def unpack(args):
+    """Helper function to convert from the format lark gives to a more useful form."""
+    (args,) = args
+    return args
```
Comment (josh146) [inline parser_utils.py:43]: This can be avoided by changing the Lark transformer options (we can have lark unpack automatically).

File: parser_utils.py (hunk 6 above)
Comment (josh146) [reply parser_utils.py:57]: Nice!

File: parser_utils.py (hunk 9 above)
Comment (josh146) [reply parser_utils.py:312]: `QDeclaration` comes from the language spec, but I agree, it's not very clear how it is useful.

**Review by josh146 — APPROVED** [review]
I'll branch off this for my PR 👍

File: parser.py (hunk 13 above)
Comment (co9olguy) [reply parser.py:255]: I asked people what they thought about this. Nobody felt as strongly as you do :laughing:

Comment (co9olguy) [reply parser.py:255]: Simultaneous gates should be applied to different wires, no? The idea of simultaneous gates is you are deliberately asking the executor to run those gates at the same time, whereas if you didn't, it would be up to the compiler to schedule

File: parser_utils.py (hunk 15 above)
Comment (co9olguy) [reply parser_utils.py:43]: That would be :100:

File: parser.py (hunk 13 above)
Comment (josh146) [reply parser.py:255]: OH! I get it. Okay, I'm now in favour, that is pretty cool.

File: parser_utils.py (hunk 14 above)
Comment (co9olguy) [reply parser_utils.py:34]: There are so many situations in the grammar that depend on the specific type of "list" you are getting. This class allows us to keep that basic information about the list type

**Comment by co9olguy** [comment]
> QDeclaration comes from the language spec, but I agree, it's not very clear how it is useful.

No it doesn't, that's something I added when hacking.

**Comment by josh146** [comment]
> No it doesn't, that's something I added when hacking.

Ah, I must have misremembered.

File: parser_utils.py (hunk 14 above)
Comment (co9olguy) [reply parser_utils.py:34]: You said ":pray: :100:" above where this was being used :laughing:

Comment (josh146) [reply parser_utils.py:34]: I didn't realize this was what the implementation looked like until I scrolled further down 😆
